### DISCUSSION TITLE
Hey Flash/Air, what did time do to you? You were so beautiful back in the days

### DISCUSSION THREAD

**Discussion by ramraver** [airsdk/Adobe-Runtime-Support#3070] [General]
Let me say this first: I guess it is safe to say that I am a member of the flash community from day one back in the days with Gabocorp, Balthaser and my company ramraver. It has been a long journey and a fantastic ride many times until, and that was the first time I hesitated  to go on sticking to that platform, Steve Jobs published the open letter stating that Flash will never get support on iOS. Many may know the real motivation for that (hint: it is called App store) For different reasons I went on with the platform anyway, mainly due to the second to none profile of Air.. But if you know Adobe a little longer, you may have similar doubts like me thinking about that company that never loved Flash, bought it to..... just kill it. Adobe has a long history of non-loyalty to dev communities and they kept their track with handing Air to Harman. Harman from that day did a fairly good job to keep it up with regular SDK updates but to me it always seemed as if the team is just to small to keep Air in the focus of a more and more mainstream headed dev community. I don't want to talk about in my eyes ridiculous things like React (+ "native" LOL) or a new release of a javascript framework happening on a daily basis. And I don't want to talk about HTML that in 2024 is not capable of other things than displaying a text or an image unformatted without the help of CSS or Javascript.

The main problem I would like to adress is that the featureset of Air is more and more limited and limiting. I made some posts here years ago about NDI support, 4K support (which we finally have), support for a video format  with alpha channel but most of it simply does not happen.

Software industry has changed a lot during the last 20 years. Did you notice that the thousands of little software companies that created nevertheless heavily used things like Winzip have just vanished? All for a simple reason - the big 5 have taken over  the entire value chain, the production chain and the marketing chain that leads to funny things like millions of people suddenly making 'macro' photo shots of bugs and bees but no one wants software from little companies that adresses real needs, sioply because the focus is on the.. yes you guess it, the big 5.

I could talk days about the journey of the last 25 years, our big successes and big failures and the reasons behind it, but it is neither the right time nor the right place. I would like to just ask Andrew, Michael and all the others being involved - where is Air heading, what is the strategic plan for it and how dou you think you can reach that goal?

**Comment by little-fat-girl** [comment]
I share your pain.  Flash died, but was never replaced. I'd love to see Harmon have a fresh vision for it, but my guess is that the subscriptions they have barely pay for the support we get on Air maintenance.

**Comment by ajwfrost** [comment]
On the Flash side .. Flash Player being intended for the browser, the problem we have is that the browser vendors have decided there's no place for it. Without those plug-in APIs, it's not possible to do the same things. There are some solutions such as the "CheerpX for Flash" which is a very clever way of keeping the plug-in working within a modern browser using a WebAssembly virtualisation layer, but the Flash Player itself isn't something that is intended as a long term solution now. And yes from a technical perspective I share a lot of your views, and it's a particular shame given the efforts Adobe put into it (with a lot of input from Google) to improve the security in the 2010s.. I think the iPhone/iPad really changed things: their popularity meant that websites had to move away from Flash in order to work on devices. To be fair to Apple, there were some issues with the stability of Flash Player on iOS (in part, and I think this was discovered too late, because of a virtual memory feature where something that happens to work on desktop Windows doesn't always work on iOS).

Anyway. Personally (and knowing the Adobe team who were involved) I would say that they actually demonstrated a bit more commitment to the development community than they needed to, in handing it over to us! They could just have stopped doing it.. instead they licensed it to us, in such a way that if Harman ever decided to stop work on it, all that we've done could be taken back by Adobe and given to someone else to continue it. The main thing that went wrong at that point was the time it took for the legal agreements, with the 64-bit requirements coming up from Android, there were no announcements about it all because everyone was waiting for it to hand over to us (and we had already done a 64-bit ARM implementation for a Flash Player version, so the core avmplus JIT code was already in place). I think a lot of people moved away from AIR at this point, which was a shame.

From our perspective: yes, the revenue we get isn't as much as we had originally planned - partly because we changed our revenue model at the very start, based on feedback from most of the people we asked (just prior to launching) - we had originally looked at a royalty-based scenario (per distributed unit) like Unity recently proposed... which was our model prior to this (and still is, for customers embedding the Flash/AIR technologies into hardware products) - but this raised a number of concerns for an application framework that people had been used to getting for free. Hence the pay-for-using-the-SDK approach, which was challenging given that we didn't have any IDE to go with it... On the encouraging side though, our revenues are at least going up slightly year on year now, and we are profitable, but we're no way near the scale of Unity so it's going to be very different I'm afraid.

In terms of strategic plan ... we have some goals and plans which are intended to improve the reach of the runtime and provide the features that people are looking for. I'm not sure how viable it is currently to reach a lot of new developers: ActionScript as a language is not something that people generally would learn, with the prevalence now of JavaScript (and maybe TypeScript). It used to be more widely used in universities for courses in animation and programming but of course they're not more focused on the web and node etc. Distriqt have had some great ideas in creating the 'apm' tool as a similar concept to Node's 'npm' and I think it would be good if more components could be developed/deployed using this - or e.g. rather than me always using "python -m http.server" when I want to run a local webserver, I could hope to at some point use "apm -run http.server" or something similar! We're also trying to improve the mechanisms for setting things up for the first time (i.e. AIR SDK Manager) as well as this aiding in trouble-shooting and support.

Some of our longer term goals:
- New platforms (we're currently investigating HarmanyOS, Nintendo Switch, and of course WebAssembly)
- Multimedia improvements are slowly creeping in
- Cross-language support i.e. compile from TypeScript to LLVM to SWF in the same way as you can from C++ (CrossBridge)
- Support updates within AIR SDK Manager (hopefully, raising/searching for bugs, submitting problem files/logs etc)

In the short term, we're also just working reactively to what people want in terms of fixes and tweaks, we're about to release the AIR 51 (pre-release version), so e.g. see https://airsdk.dev/reference/actionscript/3.0/ and click on "What's New"...

Hope that's useful info! I do value the inputs and feedback we get, and the community feel. Personally I would love to continue to do all this for the next 20 years, I like the technology and think it would be great if we can get it to the point where more and more people start to use it.

**Reply by raresn** [reply]
I was really looking forward to your response here, @ajwfrost. We've used AS since... AS2, so let's say a LONG time, and this really gives me hope! We base our livelihood on AIR right now, without any better skills that we can brag about, unfortunately, and the fact that things are really moving along is a really good thing :) 
The WebAssembly part and the cross-language support will be a **game changer** for this! 
If I can tell friends they can write TypeScript instead of AS3 and get the same results that I get with AS3... WOW!!! The WebAssembly thing means that there are plans to be able to publish for the browser again? Any insights here?

**Reply by ajwfrost** [reply]
Writing TypeScript/JavaScript using the AIR APIs and generating AIR apps is something I see as a particularly useful feature.. as it means the target audience for AIR suddenly would massively expand. But there do seem to be some challenges with the idea of TypeScript to LLVM... we've looked at a few different options here, including the possibility of just embedding V8 into AIR! 

Publishing for the web is something we had parked earlier, due to some interesting licensing-related issues, but as time goes by I think this becomes more viable. We've had a bit of advice and discussed this with the folk who developed Cheerp, CheerpJ and CheerpX and who are pretty amazing when it comes to WebAssembly! So, that's also something it would be good to look at, but we'd be doing it as an AOT-compiled platform (like iOS) as we still need to tread a little carefully around some of our license restrictions: it's got to definitely not be a replacement for the Flash Player!

Short term though: get AIR 51 published, tidy up the loose ends, and improve the multimedia stuff...

thanks

**Reply by shaucky** [reply]
Hi Andrew,
 
Yeah u mentioned an important thing: 

> ActionScript as a language is not something that people generally would learn
 
At the same time, Harman didn’t provide a convenient IDE like Unity Editor or likes, the Adobe Animate is also very not working hard now (e.g. poor support for 3D animation, bone animation (Asset Warp) through precompilation)… These phenomena are not conducive to attracting new developers to AIR, experienced developers can always come up with better solutions, while beginners may have difficulties getting started.
 
I actually reflected an issue before that the built-in AS3 compiler in Adobe Animate do not support the new operator in AIR 50. I think this issue will result in AIR not being able to work with Adobe Animate even if it supports other languages such as TypeScript. But it is already the only IDE in AIR that can be directly integrated with resources. Do we want new developers to develop in a way similar to Java? The convenient graphical advantages of AIR will be wasted.
 
If the AIR team can further collaborate with the Animate team in terms of compilation, i think it would be very worthwhile.

**Reply by ajwfrost** [reply]
I completely agree! Not having an official IDE doesn't help us.. There are a number of choices for third party IDEs of course on the programming side, and maybe that's also not particularly useful as there are too many different options!

I agree re. Animate, we are in touch with the team there and have raised some of these issues. And you're right in terms of new folk, we did tend to find that in most of the queries coming from college/uni students, they were using Animate. I'm still amazed that people continue to use Flash Builder - although the earlier versions, with the design view, I think were incredibly useful (I just found the Eclipse-based IDE very clunky...)

thanks

**Reply by raresn** [reply]
People use Flash Builder because it is easy to setup, in comparison to lets say IntelliJ which we spent 4 days or more to setup and find all the things. Right now, it will definitely take less than 1 hour to setup the very basics of a project, but for a complete noob, having to spend days just to setup things, is rough. 
Maybe a "Startup Project" could be useful? I think it would be smart. 

Since VS Code is free, maybe we can all say that that this should be official "IDE" for AIR?
A problem i found is how the AS3 plugins for IntelliJ is mentained. (and maybe VS Code?)
In that case would be cool for Harman to mentain that plugin which is written by Josh, i believe. 

"it's got to definitely not be a replacement for the Flash Player!" - tough to go around this one. What else can it be? 
Looking forward to testing the new stuff added to AIR 51, especially the OTP and websockets, plus the all the other useful addition. I belive we will imediately use over 90% of the updates there, so really cool!

**Reply by Devin-Licastro** [reply]
I would love for VS code to be official and get supported by Harman, especially if there are any easier setup guides, I got a bit overwhelmed trying to figure out how to do it all with Josh's plugin,ANEs and stuff and just gave up and stuck with FlashDevelop

Btw happy about other news in this thread like Switch and HarmonyOS, cross language support etc, good stuff!

**Reply by ajwfrost** [reply]
I was quite impressed with the asconfig set-up in Visual Studio actually, it managed to import an old Flash Builder project quite nicely for us :-)    
Yes, in VS Code the MXML/AS3 support comes from Josh/BowlerHat, it's all json-configuration based (I don't know whether VSCode supports a 'property pages' type approach, which might help with set-up?)

There's mention of this and the other IDEs at https://airsdk.dev/docs/basics/setup-an-editor

I like the idea of a start-up/sample project, samples is one of the areas I think we could definitely do with improving. The AIR Package Manager can generate/maintain the app descriptor files, but it would be also useful to step through a process like you get with doing a 'new project' in Android Studio or Xcode, to then create a placeholder AS file as well as the XML descriptor and a build configuration of your choice (asconfig, ant, moonshine I think have a similar config file, etc)

So yes, probably still some activities we can work on here!

**Reply by FliplineStudios** [reply]
> Some of our longer term goals:

> - New platforms (we're currently investigating HarmanyOS, **Nintendo Switch**, and of course WebAssembly)

I would definitely be interested in exploring Nintendo Switch builds if that were an option!  All of our games rely heavily on GPU render mode, so I guess that would be the main thing we'd be looking for if it was at all possible on the platform.

Overall though, I just really appreciate the work you all do keeping up with the frequent Apple/Google changes for mobile apps, and how responsive you are to bug reports and pushing out frequent updates.  Reporting bugs to Adobe was like shouting into the void sometimes, so it's great to see you responding to reports and discussions here and getting critical fixes out quickly!

**Reply by 2jfw** [reply]
> I completely agree! Not having an official IDE doesn't help us.. 

Not sure if this is a valid suggestion but Harman could fork, maintain and distribute their own **as3 (Flash/Flex) Plugin** for **IntelliJ IDEA**.
Source code is available here: https://github.com/JetBrains/intellij-plugins/tree/master/flex
Maybe even reaching out to Jetbrains for some kind of cooperation on this?

**Reply by lgignoud** [reply]
Very interesting idea. The plugin is mostly fine and Intellidj IDE is great, mais they don't maintain the plugin any more. The main problem for us is that we can not switch to the adl64.exe when launching the debugger, which is unfortunate; all developpments now target 64 bits platforms.

**Reply by raresn** [reply]
Another downside to intelli J is that Intellij is not free.

**Reply by 2jfw** [reply]
> Very interesting idea. The plugin is mostly fine and Intellidj IDE is great, mais they don't maintain the plugin any more. The main problem for us is that we can not switch to the adl64.exe when launching the debugger, which is unfortunate; all developpments now target 64 bits platforms.

I think they do maintain it in a way as there are regular updates when new IDEA versions are released (just check the commit history on github) but it looks like they currently are not extending it with new features (like AAB, support for the new as3 syntax like `obj.?`).

**Btw: You can simply switch to new adl64.exe by just renaming it to adl.exe!**

**Reply by ajwfrost** [reply]
I think the plug-in is maintained with a light touch .. i.e. they may not proactively be making updates, but they are willing to accept inputs from third parties e.g. see the recent change that was made here: https://github.com/JetBrains/intellij-plugins/pull/898.
But yes the cost issue for Intellij is a barrier for some. 

We do have contact with all the folk behind these IDEs/extensions (well, except for FlashDevelop - which is still my go-to tool for creating new test apps! primarily because of the simplicity in setting up a new project..)

thanks

**Reply by 2jfw** [reply]
> Btw: You can simply switch to new adl64.exe by just renaming it to adl.exe!

![adl1](https://github.com/airsdk/Adobe-Runtime-Support/assets/73781224/ca99782b-2954-44f6-8caf-720a2c456b1f)
![adl2](https://github.com/airsdk/Adobe-Runtime-Support/assets/73781224/5ac33852-33b9-432b-bfc5-40add3502442)


;-)

**Reply by 2jfw** [reply]
> But yes the cost issue for Intellij is a barrier for some.

Agreed - I mean Adobe's FlashBuilder also never was for free and used by many but a free IDE would definitely be the best

**Reply by lgignoud** [reply]
> Btw: You can simply switch to new adl64.exe by just renaming it to adl.exe!

Héhé, good point !

**Reply by 2jfw** [reply]
> Btw: You can simply switch to new adl64.exe by just renaming it to adl.exe!

@ajwfrost: Just curious, if it would make sense to rename `adl64.exe` to `adl.exe` and `adl.exe` to `adl32.exe` in upcoming AIR SDK Released because 64bit is more common today?

**Reply by shaucky** [reply]
I personally think that, any renaming behavior carries the risk of historical legacy issues…

**Reply by 2jfw** [reply]
> I personally think that, any renaming behavior carries the risk of historical legacy issues…

Windows 11 e.g. doesn't even have a 32bit version any more - it's time to move on with this and stick to today's defaults, which is 64bit

**Reply by ramraver** [reply]
No front but this seems to be an ideological answer to a fact based issue.

**Reply by ajwfrost** [reply]
We still support Windows 7 :-) 
We did actually look at switching the defaults over to 64-bit on Windows, the challenge though is that the shared AIR runtime was always a 32-bit installer, so it would add some complexity there (i.e. you can't create a "native" installer package with 64-bit, because it wouldn't work with the shared runtime). So it would impact a lot of legacy/installed apps if we tried changing this.

Potentially I guess we could switch to 64-bit for ADL, and as the default for 'bundle' creation, and just leave the native installer and shared runtime as 32-bit...

**Reply by marchbold** [reply]
I have been looking at a few options here to integrate project generation from templates etc using `apm` and then outputting configuration files for various IDEs. My initial scope was to look into vscode (due to the json nature of the configuration) and then intellij (slightly more complex but still possible). 

We have a couple of initial goals here, 
- simplify app descriptor management 
- simplify ane / swc / src lib dependency management 
- integrate changes to these automatically into project configurations

Ideally when you add a dependency or configure your project through `apm` it would automatically add it to your ide build configurations. All in early days currently but this is what we are working on to simplify this process.

**Reply by Fancy2209** [reply]
Wait Nintendo Switch? That's huge! That hasn't been possible since Scaleform (and defenitely wasn't ideal due to it beeing made for UIs)

**Comment by lgignoud** [comment]
Hi Andrew,
First thanks to your dedication to the ecosystem and the community, we're very lucky that AIR and Actionscript landed in passionnate and talented hands ! On our side, we're running a successfull business application, that went from the browser flash plugin to the desktop, thanks to the AIR SDK and Harman. We don't need new fonctionnalities or new langage syntax sugar. What we would love is improvements in AIR performances. I know Adobe worked on that back in the days, with a new compiler (AS Next or something). I know it's big task, but this will be profitable to everyone, whatever platform they use. Do you have plans to look in this area in the future ?

**Reply by ajwfrost** [reply]
Yes Adobe had been looking at "AS Next" (and "Flash Player Next" etc) back in the day! I believe they did end up publishing their thoughts/analysis about it - iirc, it was actually like an "AS4" version of ActionScript and would probably have meant an additional interpreter/vm would have been integrated alongside the AS3 (avmplus) and AS2 (actions handlers) which may have been an interesting challenge!

But we haven't really got any plans to take that forward.. there would have been some optimisations made possible via that variant of the language I think, but I'm not sure it would have been a step change. The main challenge I find with some of the business applications is that they often use Flex, which is a very top-heavy framework (check out the design of Apache Royale if you want a good example of "how to apply lessons learnt"!) so performance may not have been great even with AS4!

The one thing we had been looking at of course was ahead-of-time compilation from ActionScript to machine code, like happens on iPhone/iPad/AppleTV platforms. We've got this half-done for Android at the moment and it would be interesting to see how much of a difference it may make: talking with the Adobe folk, they weren't convinced of a massive benefit, and when we do profile some cases to see how much time is spent in the JIT verifier, yes it's not a huge amount. But it will be an interesting test case and then potentially something we can take to other (desktop) platforms too... 

thanks

**Reply by Fancy2209** [reply]
Just to confirm, AS Next was indeed AS4: https://github.com/adobe-research/ActionScript4

**Reply by hardcoremore** [reply]
Yeah performance improvements would be great. AS3 as it is, is a great language IMO but performance could be better. AOT to machine code for Android would be awesome. Thanks Andrew for all the work.

**Reply by ylazy** [reply]
I remembered many years ago AS3 is much faster compared to JS, but now it's the opposite. AVM2 is faster than AVM1, but much slower compared to JAVA or JS.

AS3
```as
var arr:Array = [];
arr[2147483646] = 2147483646; // int.MAX_VALUE - 1
var index:int = arr.indexOf(2147483646);
trace(index); // hang forever
```

JS
```js
var arr = [];
arr[2147483646] = 2147483646; // int.MAX_VALUE - 1
var index = arr.indexOf(2147483646);
console.log(index); // ~ 15 seconds to responding on my machine
```

**Comment by ramraver** [comment]
Short thoughts on some things said:
1. "Do we want new developers to develop in a way similar to Java? The convenient graphical advantages of AIR will be wasted." - 100% agreed. The high absorption rate of flash in 2000 was due to what it could do with "Multimedia" from the result perspective and the integrated development process on the development side, which united the coder and the designer into one person. True defragmentation.
2. Multimedia is and has always been a key benefit of Flash/Air. The breakthrough came with the integrated codecs of Flash Player and the use as the premiere video player on the web.
3. Expand the multimedia possibilities e.g. screen capture/virtual camera that would make it possible to write a Zoom/Teams serving app, alpha channel supporting video. yes, go video.
4. Just inspired by the fact of the undoubtedly missing flash player for the web and some words mentioned in this thread: I guess the decentralization of the app stores would possibly mean a window of opportunity for something like an "instant (air) app" that would be opened directly from within a website, providing all flash features and even more with the appropriate authorizations. Just an idea. The fragmenation of the dev process for the web is ridiculous, holding back and leading to nothing. Consolidation for this will happen, weather into this or that direction.

**Reply by ajwfrost** [reply]
Completely agree. For the first point though, we're just not in a position to create a new tool for graphics/animation. I had been wondering about generating SWFs from Adobe Express, am just trying to see whether they would add event-based animations/triggers or whether their goal is just linear-type video outputs.. but we'll continue to try to raise issues with the Animate team!

For multimedia, we're slowly improving things but I'm also hoping to get it to a point where third parties can then create and plug in different components, which then takes away some of the dependencies upon our internal team! so e.g. handling RTMP, RTSP, etc can all be written as ANEs or SWCs and just used as a source for a media pipeline. Will have to look at getting this working across a few more platforms - Android being the one we're looking at next; currently we can do basic audio and video on Windows albeit still with a lot of restrictions..

We'd seen some recent trends for mobile games to start migrating to desktop and web platforms, so yes, something we also would like to enable / capitalise on!

thanks

**Comment by Adrian-S** [comment]
I'm wondering what would the performance hit be if we would ran SWF on `<canvas>`. I just don't like the idea that JS can be decompiled, exposing the entire runtime. Still would bring back all Miniclip games.

And have you seen Lottie Web Player? Their logo looks so much as the old flash player logo, and they do animations exactly how flash did, but they use SVG and JS. And there are already many apps that do keyframe animations and transitions. Adobe Animate is still alive, but I don't know who is still using it.

**Reply by ramraver** [reply]
LOL, so true. Had the same thought when I came across that Lottie Logo and the fact of its vectorbased animation. The whole thing reminds me on the beginning of Flash, though lightweight animation is not really necessary anymore and truly a good joke if you look at the daily use of html and js on websites today that even blows up my ryzen 5950x...

**Comment by joao-cesar** [comment]
First of all, thanks, @ajwfrost and the whole Harman team for all the amazing work you have done. We're lucky that AIR is in very good hands.

Following the discussion, IMHO, the Flash technology wouldn't have been so relevant and popular without the IDE. And right now the IDE is pretty much abandoned. There is basically no feedback from bug reports and features suggestions in the prerelease forums. The latest update was very small and there is no effort in promoting the tool, although it is still very popular between young people.

AIR would have much more chances to succeed if it continues to be linked to the IDE that made it popular in first place, I think.

**Comment by Michoko92** [comment]
Thank you @ajwfrost for yours clear and thoughtful communication, and for taking care of our beloved SDK with your team. It is much appreciated.

Regarding the IDE, I can understand that some people would like a Flash IDE alternative, which includes both design and coding tools. But if we discuss at least about a code-only IDE, I'm really surprised FlashDevelop is not considered a serious possibility. Open source is always considered as the Graal in development communities, so why FlashDevelop, which could definitely be forked into AIRDevelop, could not be considered as a starting point for further development of a dedicated IDE?

Today, I'm still using FD as my main IDE. I tested a few others, even commercial ones, and I still have to find another IDE that is as fast and easy to setup. For example, the current IntelliJ IDEA implementation unzips all ANEs every time you run your code, which makes iteration much slower. Other IDEs like VS Code don't allow multiple windows layouts on different monitors (at least last time I checked). FD allows all that, you also just drag your ANEs and SWCs to your projects, and that's it. Super simple , fast and handy. There are also some nice included utilities like an SDK downloader, Shared Object analyzer... So why reinvent the wheel while FD could be a very solid foundation for future AIR-specific improvements?

I know that FD was Windows only until now, but .NET is now available for MacOS too, so maybe it would be possible to port it to Mac too? Just a thought, but I often think FD is definitely underestimated just because there was no official commits to the repo for years.

**Reply by manishpatel0404** [reply]
@Michoko92, I agree with your assessment of FlashDevelop.

**Reply by Devin-Licastro** [reply]
Also still using FD here. Feels a bit old but with some updates could be great again

**Reply by raresn** [reply]
I really think we should stay away from anything that has Flash inside it's name :) The amount of times I had to explain how AIR is not Flash is simply insane. This should be a rebranding of the system, can't have legacy stuff there.

**Reply by Michoko92** [reply]
Frankly, throwing away a very solid and working code base just because of a name would really sound absurd to me. I agree that FlashDevelop should be forked as "AIRDevelop" (like I suggested in my initial post), to avoid any confusion. But at some point, we have to admit that the goal of the IDE is to generate a SWF file, that will then be packaged by AIR -- something that FD does very well.  I do all my AIR development with FD right now.

So I guess that we won't be able to deny the origins of AIR entirely, which is basically Flash on steroids. But we can try to hide it with new names, if we are so ashamed of it.  😉 We see this kind of stuff in open source dev all the time.

**Reply by raresn** [reply]
AIR Develop sounds just great. We don't need to throw away anything, we need to market it well. 
Would love to work with a tool that is an industry "standard", like VS Code (we now use IntelliJ). All my life I felt like we were behind WITH SOMETHING because we choose the AS3 path. Thats why Typescript and up to date technologies sound super good to me for adoption purposes.
Although, Andrew should know better what to do here. 
Still, is cool to have discussion about the future of FLASH.

**Reply by Fancy2209** [reply]
[AS3 & MXML](https://github.com/BowlerHatLLC/vscode-as3mxml/wiki/Install-the-ActionScript-and-MXML-extension-for-Visual-Studio-Code) are supported by VSCode thanks to Bowler Hat's Extension

**Reply by M4rkY4** [reply]
I'm not sure if that's an optimal way.

If you look at IDE usage statistics and sum all JetBrains IDEs, they seem to have the highest popularity among IDEs. VSCode is quite close to them. So we, flash enthusiasts, have two best IDEs/IDE families covered.

It's very convenient to switch between technologies back and forth (from IDEA with Flash to either Rider with Unity or IDEA/WebStorm with Typescript or from VSCode with Flash to VSCode with Typescript and vice versa) while keeping IDE workflow identical. Moreover, it's easier to invite new people to Flash if they can use their current modern IDE and not switch to an unknown one (for them). Outdated IDE is like a bad advertisement.

If I'm not mistaken, it was planned for the Air SDK Manager to create/setup empty new project (aka boilerplate) and it looks a way better idea. It should be easier to add SharedObject reader (perhaps possible to add a flag/checkbox to recompile ANEs) to the Air SDK Manager than to port and keep FlashDevelop afloat on MacOS and Linux.

As of editor, while Animate CC is great, if anybody has enough free time, perhaps it would be possible to contribute to new PixiJS Comet Editor, which should be under active development, to be able to generate Flash/Starling content. It might improve Flash ecosystem alot. I'm not sure, though, if it's an open-source project.

**Reply by marchbold** [reply]
Thought I'd contribute some of my thoughts on this topic as I've discussed it a lot over the years. 

If there is an "official" IDE developed I believe the main reason for it should be to bring more developers to AIR.

So as @M4rkY4 mentioned IntelliJ and VSCode are two of the most commonly used IDEs by developers and would give us the easiest entry point for developers of other platforms to convert to AIR. In my daily work with native extensions I switch between multiple IDEs but always open up IntelliJ and VSCode first as they immediately cover several of the platforms I work with.  IntelliJ's flash plugin is open source and can be modified/updated, and it is fully featured providing a very functional development environment. VSCode's plugin from Josh is still in early days and very config file reliant, but VSCode is a very lightweight tool and is free. So I see both of these IDEs being very important in AIR's future and if Harman was to support an IDE I'd suggest investigating those. 

We have several priorities in terms of the tooling we are developing (there are many more but these are some of the key items relevant here):
- a way to scaffold new projects from templates (produce some basic boilerplate applications)
- generate project files to configure IDEs including injecting anes / swcs / assets into build configs (currently investigating both intellij and vscode here)
- integrating apm into vscode


Regarding Flash Develop specifically though, it was always a community driven IDE so there is no reason the community couldn't undertake maintaining it. However I wouldn't encourage Harman to engage with it. The software is specifically windows oriented and while .net ports exist on macos applications built with it have always seemed out of place to me. And more importantly I believe it presents a barrier for bringing in developers from other systems when they are forced to learn a new IDE.

**Comment by KupoGames** [comment]
Thanks for the updates! I appreciate the work that is being put into AIR, and the possibility of Switch support is cool - I'm sure that would make some more developers consider using AIR, especially those who have used Flash in the past.

AIR has allowed me to port my PC games to Android and iOS fairly easily, which is something I didn't think would have been possible back when I was still making browser Flash games. The real difficulty in mobile dev is working with the frustrating app stores, and their always-changing requirements.

For me personally, I'm happy to keep paying that AIR subscription as long as AIR continues to be stable on Windows, Android and iOS.
I would love it if AIR had hardware accelerated vector graphics on desktop, like it does on mobile. Currently Ruffle's rendering of vector graphics is many times faster than Flashplayer's and it's pretty awesome - and in a couple of years Ruffle might be a viable engine for Desktop games, though they are much further behind on mobile at the moment.

At any rate, it looks like Flash continues to live on, and I'm still happy to keep using it for large game projects.

**Reply by Fancy2209** [reply]
I believe it supports vector Tesselation in mobile by setting the renderMode to GPU (Flipline Studios does it for their mobile games) I'm not sure if it works on desktop, some say there is a difference, but the docs say it doesn't work

**Reply by KupoGames** [reply]
Yeah, I've done some tests in the past where vector animations run better on my phone than on my powerful gaming PC, so I'm pretty certain only mobile has this. :P

**Reply by Fancy2209** [reply]
Did you even have to toggle the option on?

**Reply by ajwfrost** [reply]
FYI on desktop, if you specify gpu mode, the runtime just interprets this as direct mode. I'm not entirely sure why Adobe had done this - will see if the folk we're currently in touch with have the background/reasons for it!

But I do know that the way SWF files are structured and the rendering is done, that SWF content - and SVG content likewise - would have benefitted a lot more from OpenVG rather than OpenGL. So it may just be that the performance improvements gained from using OpenGL on a desktop platform are just not as great given that there are some good optimisations for the cpu-based rendering anyway (running on desktop processors with MMX/SSE etc..)

**Reply by FliplineStudios** [reply]
> FYI on desktop, if you specify gpu mode, the runtime just interprets this as direct mode. I'm not entirely sure why Adobe had done this - will see if the folk we're currently in touch with have the background/reasons for it!

That's odd, in our tests we've actually seen a little bit of a difference between the gpu mode and direct mode on desktop, where direct handled mouse movement/sensitivity better than gpu, but also seemed to have slower performance on 4K monitors than gpu mode.  Never could figure out why there was a difference, but we stuck with gpu since it performed better than direct at 4K for our vector content.  Once in a while a player will contact us about mouse sensitivity issues (I think mainly if their screen has an odd refresh rate?), and after we have them switch the renderMode tag in the xml to direct that sometimes "fixes" it for them...

**Reply by Fancy2209** [reply]
> But I do know that the way SWF files are structured and the rendering is done, that SWF content - and SVG content likewise - would have benefitted a lot more from OpenVG rather than OpenGL. So it may just be that the performance improvements gained from using OpenGL on a desktop platform are just not as great given that there are some good optimisations for the cpu-based rendering anyway (running on desktop processors with MMX/SSE etc..)

Maybe it'd make sense to add an option for a OpenVG renderer then?

**Comment by DrNeroCF** [comment]
As much as I hate to admit it, re-releasing and updating my older games on Steam is the main thing that's keeping me in in AIR, but I'd be lying if I said reading about long-term support for AIR wasn't exciting...

I'm sure I've brought it up in every conversation I've had about Flash in the past 10 years, but I need to ask, has Harmon looked at what Scaleform was doing way back when?  Maybe AIR was doing similar things on mobile, but I'd love to the see return of the 2d stage, but software acceleration at 4k, even with today's tech, just isn't feasible, and moving everything over to Starling, while giving 2d in AIR a second wind, takes away the most unique strengths of Flash development.  Even Scaleform's code on desktop ran 10-100 times faster than Flash, it was insane to see (though on mobile the hitTesting ran far slower, especially when testing more complex vectors).

For maybe a more feasible wishlist, more control over the timeline would be life changing, like being able to start every MovieClip in a stopped state, set frame rate independently, have onEnterFrames called at a fixed tick rate, alongside display rate update events...  Though I handle all animation manually in my AS3 code, updating my older AS2 to run at 60fps has been a nightmare.

Updating the old AS2 games definitely has me missing when I could simply let an animation play without manually nextFrame()ing it at a different rate, or draw something without needing to sprite sheet it first...

**Reply by M4rkY4** [reply]
What are your thoughts on Ruffle? Looks quite promising, especially as you can embed it at your own web page. Perhaps, performance is improved as well.
Those demos on their web page look very promising.

**Reply by Fancy2209** [reply]
Having the GPU windowMode on PC would solve this issue. That wmMode tesselates the vectors, so you'd hopefully have performance similar to Scaleform GFx, but that is mobile only and if that will happen on PC is another history.
Some people claim GPU mode actually does something already on PC (who goes agaisnt the docs (though the docs are wrong sometimes), so I wonder if it compares on PC or not.
